File: code/algorithms/strings/LongestRepeatedSubstring.cpp


#include <iostream>
#include <vector>
#include <set>
#include <algorithm>

using namespace std;

class SuffixArray {
private:
    int ALPHABET_SZ = 256;
    int N;
    vector<int> T, sa, sa2, rank, c, lcp;
    vector<int> toIntArray(const string& s) {
        vector<int> text(s.length());
        for (int i = 0; i < s.length(); i++) text[i] = s[i];
        return text;
    }
    void construct() {
        int p, r;
        for (int i = 0; i < N; ++i) c[rank[i] = T[i]]++;
        for (int i = 1; i < ALPHABET_SZ; ++i) c[i] += c[i - 1];
        for (int i = N - 1; i >= 0; --i) sa[--c[T[i]]] = i;
        for (p = 1; p < N; p <<= 1) {
            for (int r = 0, i = N - p; i < N; ++i) sa2[r++] = i;
            for (int i = 0; i < N; ++i) if (sa[i] >= p) sa2[r++] = sa[i] - p;
            fill(c.begin(), c.begin() + ALPHABET_SZ, 0);
            for (int i = 0; i < N; ++i) c[rank[i]]++;
            for (int i = 1; i < ALPHABET_SZ; ++i) c[i] += c[i - 1];
            for (int i = N - 1; i >= 0; --i) sa[--c[rank[sa2[i]]]] = sa2[i];
            sa2[sa[0]] = r = 0;
            int i;
            for (i = 1; i < N; ++i) {
                if (!(rank[sa[i - 1]] == rank[sa[i]]
                    && sa[i - 1] + p < N
                    && sa[i] + p < N
                    && rank[sa[i - 1] + p] == rank[sa[i] + p])) r++;
                sa2[sa[i]] = r;
            }
            vector<int> tmp = rank;
            rank = sa2;
            sa2 = tmp;
            if (r == N - 1) break;
            ALPHABET_SZ = r + 1;
        }
    }
    void kasai() {
        lcp = vector<int>(N);
        vector<int> inv(N);
        for (int i = 0; i < N; i++) inv[sa[i]] = i;
        for (int i = 0, len = 0; i < N; i++) {
            if (inv[i] > 0) {
                int k = sa[inv[i] - 1];
                while ((i + len < N) && (k + len < N) && T[i + len] == T[k + len]) len++;
                lcp[inv[i] - 1] = len;
                if (len > 0) len--;
            }
        }
    }

public:
    SuffixArray(const string& str) : SuffixArray(toIntArray(str)) {}
    SuffixArray(const vector<int>& text) {
        T = text;
        N = text.size();
        sa = vector<int>(N);
        sa2 = vector<int>(N);
        rank = vector<int>(N);
        c = vector<int>(max(ALPHABET_SZ, N));
        construct();
        kasai();
    }
    set<string> lrs() {
        int max_len = 0;
        set<string> lrss;
        for (int i = 0; i < N; i++) {
            if (lcp[i] > 0 && lcp[i] >= max_len) {
                if (lcp[i] > max_len) lrss.clear();
                max_len = lcp[i];
                lrss.insert(string(T.begin() + sa[i], T.begin() + sa[i] + max_len));
            }
        }
        return lrss;
    }
};

int main() {
    string str1 = "ABC$BCA$CAB";
    SuffixArray sa1(str1);
    cout << "LRS(s) of " << str1 << " is/are: ";
    for (const auto& lrs : sa1.lrs()) {
        cout << lrs << " ";
    }
    cout << endl;

    string str2 = "aaaaa";
    SuffixArray sa2(str2);
    cout << "LRS(s) of " << str2 << " is/are: ";
    for (const auto& lrs : sa2.lrs()) {
        cout << lrs << " ";
    }
    cout << endl;

    string str3 = "abcde";
    SuffixArray sa3(str3);
    cout << "LRS(s) of " << str3 << " is/are: ";
    for (const auto& lrs : sa3.lrs()) {
        cout << lrs << " ";
    }
    cout << endl;

    return 0;
}

File: code/algorithms/strings/LongestRepeatedSubstring.py


class SuffixArray:
    def __init__(self, text):
        self.ALPHABET_SZ = 256
        self.N = len(text)
        self.T = text
        self.sa = [0] * self.N
        self.sa2 = [0] * self.N
        self.rank = [0] * self.N
        self.c = [0] * max(self.ALPHABET_SZ, self.N)
        self.construct()
        self.kasai()

    def construct(self):
        p = r = 0
        for i in range(self.N):
            self.c[self.rank[i] == self.T[i]] += 1
        for i in range(1, self.ALPHABET_SZ):
            self.c[i] += self.c[i - 1]
        for i in range(self.N - 1, -1, -1):
            self.sa[self.c[self.T[i]] - 1] = i
            self.c[self.T[i]] -= 1
        p = 1
        while True:
            r = 0
            for i in range(self.N - p, self.N):
                self.sa2[r] = i
                r += 1
            for i in range(self.N):
                if self.sa[i] >= p:
                    self.sa2[r] = self.sa[i] - p
                    r += 1
            self.c = [0] * max(self.ALPHABET_SZ, self.N)
            for i in range(self.N):
                self.c[self.rank[i]] += 1
            for i in range(1, self.ALPHABET_SZ):
                self.c[i] += self.c[i - 1]
            for i in range(self.N - 1, -1, -1):
                self.sa[self.c[self.rank[self.sa2[i]]] - 1] = self.sa2[i]
                self.c[self.rank[self.sa2[i]]] -= 1
            self.sa2[self.sa[0]] = r = 0
            for i in range(1, self.N):
                if not (self.rank[self.sa[i - 1]] == self.rank[self.sa[i]] and self.sa[i - 1] + p < self.N and self.sa[i] + p < self.N and self.rank[self.sa[i - 1] + p] == self.rank[self.sa[i] + p]):
                    r += 1
                self.sa2[self.sa[i]] = r
            self.tmp = self.rank
            self.rank = self.sa2
            self.sa2 = self.tmp
            if r == self.N - 1:
                break
            self.ALPHABET_SZ = r + 1
            p <<= 1

    def kasai(self):
        self.lcp = [0] * self.N
        inv = [0] * self.N
        for i in range(self.N):
            inv[self.sa[i]] = i
        i = 0
        while i < self.N:
            if inv[i] > 0:
                k = self.sa[inv[i] - 1]
                length = 0
                while i + length < self.N and k + length < self.N and self.T[i + length] == self.T[k + length]:
                    length += 1
                self.lcp[inv[i] - 1] = length
                if length > 0:
                    length -= 1
            i += 1

    def lrs(self):
        max_len = 0
        lrss = set()
        for i in range(self.N):
            if self.lcp[i] > 0 and self.lcp[i] >= max_len:
                if self.lcp[i] > max_len:
                    lrss.clear()
                max_len = self.lcp[i]
                lrss.add("".join(map(chr, self.T[self.sa[i]:self.sa[i] + max_len])))
        return lrss


def main():
    str1 = "ABC$BCA$CAB"
    sa1 = SuffixArray(list(map(ord, str1)))
    print("LRS(s) of", str1, "is/are:", sa1.lrs())

    str2 = "aaaaa"
    sa2 = SuffixArray(list(map(ord, str2)))
    print("LRS(s) of", str2, "is/are:", sa2.lrs())

    str3 = "abcde"
    sa3 = SuffixArray(list(map(ord, str3)))
    print("LRS(s) of", str3, "is/are:", sa3.lrs())


if __name__ == "__main__":
    main()
